File: Java/3_Longest_Substring_Without_Repeating_Characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int length = s.length();
        Map<Character, Integer> charMap = new HashMap<>();
        int substrBegin = 0, maxLength = 0;

        for (int curr = 0; curr < length; curr++) {
            char currChar = s.charAt(curr);
            if (!charMap.containsKey(currChar) || charMap.get(currChar) < substrBegin){
                charMap.put(currChar, curr);
                maxLength = Math.max(maxLength, curr - substrBegin + 1);
            }
            else {
                substrBegin = charMap.get(currChar) + 1;
                charMap.put(currChar, curr);
            }
        }
        return maxLength;
    }
}

File: C++/3_Longest_Substring_Without_Repeating_Characters.cpp
#include <iostream>
#include <map>
using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int length = s.length();
        int maxLength = 0;
        unordered_map<char, int> charMap;
        int substrBegin = 0;

        for (int currchar = 0; currchar < length; currchar++) {
            if(charMap.count(s[currchar]) == 0 or charMap[s[currchar]] < substrBegin){
                charMap[s[currchar]] = currchar;
                maxLength = max(maxLength, currchar - substrBegin + 1);
            }
            else {
                substrBegin = charMap[s[currchar]] + 1;
                charMap[s[currchar]] = currchar;
            }
        }

        return maxLength; 
    }
};
